src/sidebar/util/test/group-organizations-test.js

```diff
@@ -0,0 +1,174 @@
+        { name: 'Beryllium', organization: org2},
+        { name: 'Butane', organization: org2 },
+        { name: 'Cadmium', organization: org},
+      ];
+
+      const sortedGroups = groupsByOrganization(groups);
+
+      assert(typeof sortedGroups[0].logo === 'string');
+      assert(typeof sortedGroups[1].logo === 'undefined');
+      assert(typeof sortedGroups[2].logo === 'string');
+      assert(typeof sortedGroups[3].logo === 'undefined');
+
+    });
+  });
+
+  context('when encountering missing data', function () {
+
+    it('should be able to sort without any groups in the default org', function () {
+      const org = orgFixtures.organization({ name:'Aluminum' });
+      const groups = [
+        { name: 'Aluminum', organization: org },
+        { name: 'Beryllium', organization: org},
+        { name: 'Cadmium', organization: org},
+      ];
+
+      const sortedGroups = groupsByOrganization(groups);
+
+      sortedGroups.forEach((group, idx) => {
+        assert.equal(group.name, groups[idx].name);
+      });
+
+    });
+
+    it('should omit any groups without an organization', function () {
+      const org = orgFixtures.organization({ name:'Europium' });
+      const groups = [
+        { name: 'Aluminum', organization: org },
+        { name: 'Beryllium', organization: org},
+        { name: 'Butane' },
+        { name: 'Cadmium', organization: org},
+      ];
+
+      const sortedGroups = groupsByOrganization(groups);
+
+      assert.equal(sortedGroups.length, groups.length - 1);
+      sortedGroups.forEach(group => {
+        assert.notEqual(group.name, 'Butane');
+      });
+    });
+
+    it('should omit any groups with unexpanded organizations', function () {
+      const org = orgFixtures.organization({ name:'Europium' });
+      const groups = [
+        { name: 'Aluminum', organization: org },
+        { name: 'Beryllium', organization: org},
+        { name: 'Butane', organization: 'foobar' },
+        { name: 'Cadmium', organization: org},
+      ];
+
+      const sortedGroups = groupsByOrganization(groups);
+
+      assert.equal(sortedGroups.length, groups.length - 1);
+      sortedGroups.forEach(group => {
+        assert.notEqual(group.name, 'Butane');
+      });
+
+    });
+
+    it ('should omit logo property if not present on organization', function () {
+      const org = orgFixtures.organization({ logo: undefined });
+      const org2 = orgFixtures.organization({ logo: null });
+      const groups = [
+        { name: 'Aluminum', organization: org },
+        { name: 'Beryllium', organization: org2},
+        { name: 'Butane', organization: org2 },
+        { name: 'Cadmium', organization: org},
+      ];
+
+      const sortedGroups = groupsByOrganization(groups);
+
+      sortedGroups.forEach(group => {
+        assert(typeof group.logo === 'undefined');
+      });
+    });
+  });
+
+  context('when building data structures', function () {
+
+    it ('should deep-clone group objects', function () {
```

Comment:
The main invariant you want to ensure is that the groups passed to `groupsByOrganization` don't get modified. One way to do this is to use the `seamless-immutable` lib to catch accidental mutations (it has no effect in production):

```js
const immutable = require('seamless-immutable');

function group() {
   const group = { ... };
   return immutable(group);
}
```

Search for references to this library for some existing examples in other tests. Otherwise it shouldn't matter whether the code does a shallow-clone or a deep-clone.

Reply:
Ahhh! Thank you!

Reply:
@robertknight Is there a way to write tests for this — that is, will an immutable Throw in the test environment if you try to futz with it?

Reply:
Yes it will. It deep-freezes objects/arrays in dev environments and does nothing in production.